File: app/src/main/java/br/ufpa/smartufpa/fragments/AddBuildingFragment.java
package br.ufpa.smartufpa.fragments;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputEditText;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.Spinner;

import org.json.JSONException;
import org.json.JSONObject;

import br.ufpa.smartufpa.R;
import br.ufpa.smartufpa.activities.SelectCategoryActivity;

public class AddBuildingFragment extends NewPlaceFragment {

    public static final String TAG = AddBuildingFragment.class.getName();

    // Views
    private TextInputEditText edtxtName;
    private TextInputEditText edtxtDescription;
    private TextInputEditText edtxtWebsite;
    private TextInputEditText edtxtOpeningTime;
    private Spinner spinner;
    private TextInputEditText edtxtClosingTime;
    private Button btnConfirm;
    private Button btnCancel;

    // Keys for JSON Object
    private final String keyName = "name";
    private final String keyDescription = "description";
    private final String keyWebsite = "website";
    private final String keyOpeningTime = "opening_time";
    private final String keyClosingTime = "closing_time";
    private final String keyAdministrativeRole = "administrative_role";

    private double latitude;
    private double longitude;

    private onNewPlaceListener listener;


    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        final View view = inflater.inflate(R.layout.fragment_add_building, container, false);
        edtxtName = view.findViewById(R.id.input_bld_name);
        edtxtDescription = view.findViewById(R.id.input_bld_description);
        edtxtWebsite = view.findViewById(R.id.input_bld_website);
        edtxtOpeningTime = view.findViewById(R.id.input_bld_openingtime);
        edtxtClosingTime = view.findViewById(R.id.input_bld_closingtime);
        spinner = view.findViewById(R.id.spin_bld_administrative_role);

        ArrayAdapter<CharSequence> adapter = ArrayAdapter.createFromResource(view.getContext(),
                R.array.spinner_administrative_role, android.R.layout.simple_spinner_item);
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spinner.setAdapter(adapter);

        btnConfirm = view.findViewById(R.id.btn_bld_confirm);
        btnCancel = view.findViewById(R.id.btn_bld_cancel);

        final View.OnClickListener onClickListener = new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                switch (view.getId()) {
                    case R.id.btn_bld_confirm:
                        updateJson();
                        break;
                    case R.id.btn_bld_cancel:
                       getActivity().onBackPressed();
                        break;
                }
            }
        };

        btnConfirm.setOnClickListener(onClickListener);
        btnCancel.setOnClickListener(onClickListener);

        edtxtOpeningTime.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(final View view) {
                showTimePickerDialog(edtxtOpeningTime);
            }
        });

        edtxtClosingTime.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(final View view) {
                showTimePickerDialog(edtxtClosingTime);
            }
        });



        return view;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        Bundle bundle = getArguments();
        latitude = bundle != null ? bundle.getDouble(SelectCategoryActivity.KEY_LATITUDE) : 0;
        longitude = bundle != null ? bundle.getDouble(SelectCategoryActivity.KEY_LONGITUDE) : 0;
    }


    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if(context instanceof onNewPlaceListener){
            listener = (onNewPlaceListener) context;
        }else{
            throw new ClassCastException();
        }
    }

    @Override
    public void updateJson() {
        JSONObject json = new JSONObject();
        try {
            json.put(keyName, edtxtName.getText().toString());
            json.put(SelectCategoryActivity.KEY_LATITUDE, latitude);
            json.put(SelectCategoryActivity.KEY_LONGITUDE, longitude);
            json.put(keyDescription, edtxtDescription.getText().toString());
            json.put(keyWebsite, edtxtWebsite.getText().toString());
            json.put(keyOpeningTime, edtxtOpeningTime.getText().toString());
            json.put(keyClosingTime, edtxtClosingTime.getText().toString());

            if (spinner.getSelectedItemId() == 0)
                json.put(keyAdministrativeRole,"");
            else
                json.put(keyAdministrativeRole, spinner.getSelectedItem().toString());
        } catch (JSONException e) {
            e.printStackTrace();
        }
        listener.getFormJSON(json);
    }


}

File: app/src/main/java/br/ufpa/smartufpa/fragments/AddBuildingFragment.kt
package br.ufpa.smartufpa.fragments

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.Toast

import br.ufpa.smartufpa.R
import br.ufpa.smartufpa.activities.SelectCategoryActivity
import br.ufpa.smartufpa.models.smartufpa.APIResponse
import br.ufpa.smartufpa.models.smartufpa.Building
import br.ufpa.smartufpa.utils.retrofit.PlaceAPI
import br.ufpa.smartufpa.utils.retrofit.RetrofitClient
import kotlinx.android.synthetic.main.fragment_add_building.*
import kotlinx.android.synthetic.main.fragment_add_building.view.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit

class AddBuildingFragment : NewPlaceFragment() {


//    // Keys for JSON Object
//    private val keyName = "name"
//    private val keyDescription = "description"
//    private val keyWebsite = "website"
//    private val keyOpeningTime = "opening_time"
//    private val keyClosingTime = "closing_time"
//    private val keyAdministrativeRole = "administrative_role"

    companion object {

        val TAG = AddBuildingFragment::class.java.name
        private lateinit var retrofitClient : Retrofit
        private lateinit var placeAPIClient : PlaceAPI

        @JvmStatic
        fun newInstance(context: Context?): AddBuildingFragment {
            retrofitClient = RetrofitClient.getInstance(context, null)
            placeAPIClient = retrofitClient.create(PlaceAPI::class.java)
            return AddBuildingFragment()
        }
    }

    private var latitude: Double = 0.toDouble()
    private var longitude: Double = 0.toDouble()

    private var listener: NewPlaceFragment.onNewPlaceListener? = null




    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
      return inflater.inflate(R.layout.fragment_add_building, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val adapter = ArrayAdapter.createFromResource(view.context,
                R.array.spinner_administrative_role, android.R.layout.simple_spinner_item)
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        spin_bld_administrative_role!!.adapter = adapter

        val onClickListener = View.OnClickListener { view ->
            when (view.id) {
                R.id.btn_bld_confirm -> sendToServer()
                R.id.btn_bld_cancel -> activity!!.onBackPressed()
            }
        }

        btn_bld_confirm!!.setOnClickListener(onClickListener)
        btn_bld_cancel!!.setOnClickListener(onClickListener)

        input_bld_openingtime!!.setOnClickListener { showTimePickerDialog(input_bld_openingtime) }

        input_bld_closingtime!!.setOnClickListener { showTimePickerDialog(input_bld_closingtime) }


    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        val bundle = arguments
        latitude = bundle?.getDouble(SelectCategoryActivity.KEY_LATITUDE) ?: 0.0
        longitude = bundle?.getDouble(SelectCategoryActivity.KEY_LONGITUDE) ?: 0.0
    }


    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (context is NewPlaceFragment.onNewPlaceListener) {
            listener = context
        } else {
            throw ClassCastException()
        }
    }

    fun sendToServer(){
        val building = Building(0,latitude, longitude, input_bld_name.text.toString(),
                            "","","")
        val call: Call<APIResponse> = placeAPIClient.addBuilding(building)
        call.enqueue(object : Callback<APIResponse> {
            override fun onResponse(call: Call<APIResponse>, response: Response<APIResponse>) {
                val status = response.isSuccessful
                if(status){
                    Toast.makeText(context,response.message(),Toast.LENGTH_LONG).show()
                    activity?.onBackPressed()
                }
            }

            override fun onFailure(call: Call<APIResponse>, t: Throwable) {
                t.printStackTrace()
            }
        });
    }

    override fun updateJson() {
//        val json = JSONObject()
//        try {
//            json.put(keyName, edtxtName!!.text.toString())
//            json.put(SelectCategoryActivity.KEY_LATITUDE, latitude)
//            json.put(SelectCategoryActivity.KEY_LONGITUDE, longitude)
//            json.put(keyDescription, edtxtDescription!!.text.toString())
//            json.put(keyWebsite, edtxtWebsite!!.text.toString())
//            json.put(keyOpeningTime, edtxtOpeningTime!!.text.toString())
//            json.put(keyClosingTime, edtxtClosingTime!!.text.toString())
//
//            if (spinner!!.selectedItemId == 0L)
//                json.put(keyAdministrativeRole, "")
//            else
//                json.put(keyAdministrativeRole, spinner!!.selectedItem.toString())
//        } catch (e: JSONException) {
//            e.printStackTrace()
//        }
//
//        listener!!.getFormJSON(json)
    }






}
